### PR TITLE
DAOS-623 build: Use pip to install deps on Ubuntu.

### PR DESCRIPTION
**Pull request by ashleypittman**
Install python dependencies rather than OS ones so we get to
control the versions used.

This allows daos to build in Docker under ubuntu:22.04 as it avoids
the the meson/spdk build issue.

Resolve issues with the utils/scripts/ files and update the documentation on their use.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
This brings Ubuntu dockerfile inline with Rocky 8 and allows daos to be compiled on Ubuntu 22:04 although it does not change what CI/GHA uses.

**Review by jolivier23 — DISMISSED** [review]

**Review by johannlombardi — CHANGES_REQUESTED** [review]
Well, the way i see this patch is that it actually uses a different way to install meson for Ubuntu compared to other distributions. We should either use pip everywhere and add it to the documentation as well or use the distrib packages and fix the issue in SPDK. I am not keen on having a different path just for Ubuntu.

**Comment by ashleypittman** [comment]
> Well, the way i see this patch is that it actually uses a different way to install meson for Ubuntu compared to other distributions. We should either use pip everywhere and add it to the documentation as well or use the distrib packages and fix the issue in SPDK. I am not keen on having a different path just for Ubuntu.

That's not quite the case.  We've long used pip in individual dockerfiles, mostly either for packages not in the distro like pyelftools or more recently scons to avoid the EL8 bug.  Recently I've been trying to use virtual environments and ensure that our tools can work with them, https://github.com/daos-stack/daos/pull/8590 was a recent change the the build to allow their use and migrate the EL8 dockerfile so this PR really brings Ubuntu in-line with Rocky.  I guess I should have updated the documentation at the same time, and migrate leap over as part of this PR?

**Review by johannlombardi — APPROVED** [review]

**Comment by ashleypittman** [comment]
I'm still trying to get this to work in CI as it appears Jenkins does not honour PATH settings from the dockerfile, and it should probably have https://github.com/daos-stack/daos/pull/8796 rolled into it which is removing some regressions from the artifactory change.

**Review by johannlombardi — APPROVED** [review]

**Review by brianjmurrell — DISMISSED** [review]
Expanding the risk profile of using unsigned code pulled frequently and randomly from pypi doesn't seem like a good idea to me.  Certainly when there are packages for those modules available from distros which are maintained by the distros.

Indeed, I would have raised this concern on #8590 had I seen it before it landed.

File: utils/scripts/install-centos7.sh
Comment (ashleypittman) [reply utils/scripts/install-centos7.sh:67]: This is for build, not deployment, and we're aiming to increase the coverage and flexibility of what we can test.  This PR also adds the BASE_DISTRO arg that already exists in el8 and puts it in leap and ubuntu so as a separate activity outside of CI so far I'm been using these changes to verify that we can run with alternative distributions and keep up-to-date with upstream.

Certainly for Ubuntu the distro packages are not capable of building daos without using pypi.  The same is true for scons on el8 where this change is already applied and centos 7 which on current master is already calling pip without referencing the requirments.txt file.

I believe the Ubuntu 22.04 update is blocked on this (i.e. it won't proceed without it) and these changes should also assist in the Leap 15.4 update, at least in terms of the Github actions configuration.

Comment (brianjmurrell) [reply utils/scripts/install-centos7.sh:67]: > Certainly for Ubuntu the distro packages are not capable of building daos without using pypi.

Yes, it's understood that some (hopefully minimal (as can be)) use of pypi might be necessary as we just don't have the bandwidth to roll the required RPMs.  But I would argue that there should at least be tickets for that minimal set of pypi modules that we do use as our ideal working scenario is to use RPMs, presumably maintained by a human and validated  -- whether those are from upstream or by ourselves.  If there are no tickets, the work will never get prioritized and scheduled.

> The same is true for scons on el8

Not true: https://github.com/daos-stack/scons/tree/master

> I believe the Ubuntu 22.04 update is blocked on this (i.e. it won't proceed without it) and these changes should also assist in the Leap 15.4 update, at least in terms of the Github actions configuration.

Why would this assist on Leap 15.4?  One would presume that python module RPMs would be the most up-to-date there.

**Review by jolivier23 — APPROVED** [review]

**Comment by johannlombardi** [comment]
IIRC, this is only for building in docker and for folks who want to do local build, right? This has no impact on how we generate our RPMs. Correct? It sounds like reducing the number of RPM packages to be installed (usually a request to your sysadmin) is a good idea for this path. @brianjmurrell, do you disagree?

**Review by johannlombardi — APPROVED** [review]

**Comment by johannlombardi** [comment]
Ping.

**Comment by brianjmurrell** [comment]
> IIRC, this is only for building in docker and for folks who want to do local build, right?

Right.  But we are effectively providing to people a way of building that ls "less secure".

> This has no impact on how we generate our RPMs. Correct?

Correct.

> It sounds like reducing the number of RPM packages to be installed (usually a request to your sysadmin) is a good idea for this path. @brianjmurrell, do you disagree?

I would agree if we could find a way that was safer than pulling source code from a completely unauthenticated (i.e. not even cryptographically authenticated) and actively hacked (i.e. typo-squatting in the absolute least) source.

But given that the exposure would just be local I can disagree and commit here.

**Pull request merged**
